rpcs3/Emu/Audio/FAudio/faudio_enumerator.cpp

```diff
@@ -0,0 +1,95 @@
+#ifndef HAVE_FAUDIO
+#error "FAudio support disabled but still being built."
+#endif
+
+#include "Emu/Audio/FAudio/faudio_enumerator.h"
+#include <array>
+#include <algorithm>
+#include "Utilities/StrUtil.h"
+#include "util/logs.hpp"
+
+LOG_CHANNEL(faudio_dev_enum);
+
+faudio_enumerator::faudio_enumerator() : audio_device_enumerator()
+{
+	FAudio *tmp{};
+
+	if (u32 res = FAudioCreate(&tmp, 0, FAUDIO_DEFAULT_PROCESSOR))
+	{
+		faudio_dev_enum.error("FAudioCreate() failed(0x%08x)", res);
+		return;
+	}
+
+	// All succeeded, "commit"
+	instance = tmp;
+}
+
+faudio_enumerator::~faudio_enumerator()
+{
+	if (instance != nullptr)
+	{
+		FAudio_StopEngine(instance);
+		FAudio_Release(instance);
+	}
+}
+
+std::vector<audio_device_enumerator::audio_device> faudio_enumerator::get_output_devices()
+{
+	if (!instance)
+	{
+		return {};
+	}
+
+	u32 dev_cnt{};
+	if (u32 res = FAudio_GetDeviceCount(instance, &dev_cnt))
+	{
+		faudio_dev_enum.error("FAudio_GetDeviceCount() failed(0x%08x)", res);
+		return {};
+	}
+
+	if (dev_cnt == 0)
+	{
+		faudio_dev_enum.warning("No devices available");
+		return {};
+	}
+
+	std::vector<audio_device> device_list{};
+
+	for (u32 dev_idx = 0; dev_idx < dev_cnt; dev_idx++)
+	{
+		FAudioDeviceDetails dev_info{};
+
+		if (u32 res = FAudio_GetDeviceDetails(instance, dev_idx, &dev_info))
+		{
+			faudio_dev_enum.error("FAudio_GetDeviceDetails() failed(0x%08x)", res);
+			continue;
+		}
+
+		if (std::char_traits<wchar_t>::length(std::bit_cast<wchar_t*>(&dev_info.DisplayName[0])) == 0)
+		{
+			continue;
+		}
+
+		audio_device dev =
+		{
+			.id = std::to_string(dev_idx),
```

Comment:
I was wondering why the deviceId isn't used here.

Reply:
Same thing in faudio

Reply:
What if the devices are changed.
Won't this map output to the wrong device if the config entry is let's say 2?

Reply:
I don't see a better way

Reply:
You could probably use the name instead as well

Reply:
Name may not be unique

Reply:
I guess it doesn't matter then